### PR TITLE
[SYCL] Add the notion of default contexts.

### PR DESCRIPTION
**Pull request by jbrodman**
A default context is now defined per platform.  The default context initially contains all devices in the platform.
Queue creation will use the device's platform's default context instead of creating a new context with that device.

The new extension introduces `ext_oneapi_get_default_context()` member function to the platform class.

DPC++ runtime allows control of this feature with SYCL_ENABLE_DEFAULT_CONTEXTS env variable. It is enabled by default on Linux and disabled on Windows.

### CODE REVIEW

File: sycl/test/on-device/host-interop-task/interop-task.cpp [hunk 1]
```diff
@@ -118,15 +118,16 @@ void test2() {
   buffer<int, 1> Buffer1{BUFFER_SIZE};
   buffer<int, 1> Buffer2{BUFFER_SIZE};
 
+  auto Device = default_selector().select_device();
   // init the buffer with a'priori invalid data
   {
-    queue Q;
+    queue Q(context(Device), Device);
```
Comment (alexbatashev) [inline sycl/test/on-device/host-interop-task/interop-task.cpp:124]: Why is this necessary? SYCL spec says, that default constructor should construct a queue using `default_selector`. Is that behavior changed? Should `sycl::queue` docs be updated to reflect it then?

File: sycl/source/detail/platform_impl.hpp [hunk 2]
```diff
@@ -191,12 +192,18 @@ class platform_impl {
...
+  void pushDefaultContext(context Context);
+  void popDefaultContext();
+
 private:
   bool MHostPlatform = false;
   RT::PiPlatform MPlatform = 0;
   std::shared_ptr<plugin> MPlugin;
   std::vector<std::weak_ptr<device_impl>> MDeviceCache;
   std::mutex MDeviceMapMutex;
+  std::vector<context> MDefaultContextStack;
```
Comment (alexbatashev) [inline sycl/source/detail/platform_impl.hpp:206]: That's a bit confusing. Variable name says it's a stack, but its usages indicate it's a vector. Why not use `std::stack`?

```suggestion
  std::stack<context> MDefaultContextStack;
```

File: sycl/test/on-device/host-interop-task/interop-task.cpp (hunk 1 above)
Comment (jbrodman) [reply sycl/test/on-device/host-interop-task/interop-task.cpp:124]: I found a bug related to the interop handle and contexts unrelated to this change.  This test change preserves the original behavior.

File: sycl/source/detail/platform_impl.hpp (hunk 2 above)
Comment (jbrodman) [reply sycl/source/detail/platform_impl.hpp:206]: Probably just remove Stack from the name.  Choice of container isn't really important.

File: sycl/test/on-device/host-interop-task/interop-task.cpp (hunk 1 above)
Comment (alexbatashev) [reply sycl/test/on-device/host-interop-task/interop-task.cpp:124]: Ok

File: sycl/source/detail/platform_impl.hpp (hunk 2 above)
Comment (alexbatashev) [reply sycl/source/detail/platform_impl.hpp:206]: Yes, that's also an options

File: sycl/test/on-device/defaultcontext/default.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t.out
```
Comment (olegmaslovatintel) [inline sycl/test/on-device/defaultcontext/default.cpp:1]: Shouldn't  the on-device tests be moved to https://github.com/intel/llvm-test-suite? There are contained E2E tests. please contact @vladimirlaz  on the details

Comment (jbrodman) [reply sycl/test/on-device/defaultcontext/default.cpp:1]: My understanding was that they go here first.

**Review by alexbatashev — APPROVED** [review]
LGTM

**Comment by romanovvlad** [comment]
@jbrodman Could you please point to the extension document?

File: sycl/doc/extensions/PlatformContext/PlatformContext.adoc [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+:source-highlighter: pygments
+:icons: font
+:y: icon:check[role="green"]
+:n: icon:times[role="red"]
+
+== Platform Default Contexts
+
+This extension adds the notion of a default SYCL context per SYCL platform. The default context for each platform contains all devices in the platform.
+
+The platform class gains three new methods: `get_default_context`, `push_default_context`, and `pop_default_context`.
```
Comment (romanovvlad) [inline sycl/doc/extensions/PlatformContext/PlatformContext.adoc:13]: It would be nice if functions are described in a table similar to how it's done in other extensions, like https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/ParallelForSimplification/SYCL_INTEL_parallel_for_simplification.asciidoc. Otherwise, at least, their signature is not clear.

I'm OK if it is done in a separate PR.

File: sycl/include/CL/sycl/platform.hpp [hunk 5]
```diff
@@ -126,6 +126,22 @@ class __SYCL_EXPORT platform {
...
 
+  /// Return this platform's default context
+  ///
+  /// \return the default context
+  context get_default_context() const;
+
+  /// Pushes a new default context
+  ///
+  /// \param the new default context
+  void push_default_context(context Context);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/platform.hpp:137]:
```suggestion
  void push_default_context(const context &Context);
```
Can we change it to take a reference?

File: sycl/source/detail/platform_impl.cpp [hunk 6]
```diff
@@ -389,6 +390,31 @@ bool platform_impl::has(aspect Aspect) const {
 #include <CL/sycl/info/platform_traits.def>
 #undef __SYCL_PARAM_TRAITS_SPEC
 
+context platform_impl::getDefaultContext() {
+  if (!MDefaultContexts.empty())
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.cpp:394]: It seems it can be a source of races. For example: one thread enters the if while another thread removes a context in the `popDefaultContext` method. Could you please fix that?

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/source/detail/queue_impl.hpp [hunk 7]
```diff
@@ -59,11 +59,7 @@ class queue_impl {
              const property_list &PropList)
       : queue_impl(Device,
                    detail::getSyclObjImpl(
-                       context(createSyclObjFromImpl<device>(Device), {},
```
Comment (romanovvlad) [inline sycl/source/detail/queue_impl.hpp:62]: It seems the patch removes support for property::context::cuda.  Could you please restore it? 

Probably this property should be changed to be a property of context. So far, we probably could create a new context if use_primary_context is passed.

File: sycl/doc/extensions/PlatformContext/PlatformContext.adoc [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+| Member Function | Description
+
+|
+[source,c++]
+----
+context get_default_context()
+----
+
+| Returns the current default context for this `platform`
+
```
Comment (romanovvlad) [inline sycl/doc/extensions/PlatformContext/PlatformContext.adoc:26]: It would be nice if the extension is mentioned in the https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/README.md

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/source/detail/platform_impl.hpp [hunk 9]
```diff
@@ -180,12 +182,16 @@ class platform_impl {
...
 
+  context getDefaultContext();
+
 private:
   bool MHostPlatform = false;
   RT::PiPlatform MPlatform = 0;
   std::shared_ptr<plugin> MPlugin;
   std::vector<std::weak_ptr<device_impl>> MDeviceCache;
   std::mutex MDeviceMapMutex;
+  std::vector<context> MDefaultContexts;
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.hpp:193]: Since the doc now says that there is one default context per platform, does it make sense to have just a value here instead of `std::vector`?

File: sycl/source/detail/platform_impl.cpp [hunk 10]
```diff
@@ -389,6 +390,19 @@ bool platform_impl::has(aspect Aspect) const {
...
+  const std::lock_guard<std::mutex> Guard(MContextMutex);
+
+  if (!MDefaultContexts.empty())
+    return MDefaultContexts.back();
+  else {
+    auto Devices = get_devices();
+    MDefaultContexts.emplace_back(Devices);
+
+    return MDefaultContexts.back();
+  }
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.cpp:403]:
```suggestion
  auto Devices = get_devices();
  MDefaultContexts.emplace_back(Devices);

  return MDefaultContexts.back();
```
Suggest avoiding explicit else branch because if body has return statement.

File: sycl/source/detail/platform_impl.hpp [hunk 11]
```diff
@@ -180,12 +182,16 @@ class platform_impl {
...
+  std::mutex MContextMutex;
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.hpp:194]:
```suggestion
  std::mutex MDefaultContextMutex;
```

File: sycl/test/on-device/defaultcontext/default.cpp [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+int main() {
+  sycl::queue q;
+  sycl::queue q2;
+
+  auto c1 = q.get_context();
+  auto c2 = q2.get_context();
+
+  auto d1 = q.get_device();
+  auto d2 = q2.get_device();
+
```
Comment (romanovvlad) [inline sycl/test/on-device/defaultcontext/default.cpp:25]: Please, add checks for `platform::get_default_context`.

File: sycl/test/abi/sycl_symbols_linux.dump [hunk 13]
```diff
@@ -5,7 +5,6 @@
 
 # RUN: env LLVM_BIN_PATH=%llvm_build_bin_dir python %sycl_tools_src_dir/abi_check.py --mode check_symbols --reference %s %sycl_libs_dir/libsycl.so
 # REQUIRES: linux
-# UNSUPPORTED: libcxx
```
Comment (romanovvlad) [inline sycl/test/abi/sycl_symbols_linux.dump:8]: This line needs to be restored.

File: sycl/doc/extensions/PlatformContext/PlatformContext.adoc [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+James Brodman <james.brodman@intel.com>
+v0.1
+:source-highlighter: pygments
+:icons: font
+:y: icon:check[role="green"]
+:n: icon:times[role="red"]
+
+== Platform Default Contexts
+
+This extension adds the notion of a default SYCL context per SYCL platform. The default context for each platform contains all devices in the platform.
```
Comment (romanovvlad) [inline sycl/doc/extensions/PlatformContext/PlatformContext.adoc:11]: It seems that the lifetime of the default context is not clear.
It can be important in situations, like: SYCL2020 says that in some cases unconsumed async exceptions should be passed to the `context` async handler during the context destruction.

File: sycl/source/detail/queue_impl.hpp (hunk 7 above)
Comment (jbrodman) [reply sycl/source/detail/queue_impl.hpp:62]: So I can't find where this "if" could possibly be true.

As far as I can tell, the only two mentions of DefaultContextType are in this file, and it's constexpr assigned to a different value than the one checked in the if.

File: sycl/source/detail/platform_impl.hpp (hunk 9 above)
Comment (jbrodman) [reply sycl/source/detail/platform_impl.hpp:193]: Yes

File: sycl/source/detail/platform_impl.cpp (hunk 10 above)
Comment (jbrodman) [reply sycl/source/detail/platform_impl.cpp:403]: 'k

File: sycl/source/detail/platform_impl.hpp (hunk 11 above)
Comment (jbrodman) [reply sycl/source/detail/platform_impl.hpp:194]: cool

File: sycl/source/detail/platform_impl.cpp [hunk 15]
```diff
@@ -391,16 +391,16 @@ bool platform_impl::has(aspect Aspect) const {
...
-  else {
-    auto Devices = get_devices();
-    MDefaultContexts.emplace_back(Devices);
+  if (!MDefaultContext)
+    return *MDefaultContext;
 
-    return MDefaultContexts.back();
-  }
+  // Lazily instantiate default context
+  auto Devices = get_devices();
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.cpp:400]: Please, do not use auto here.

Comment (jbrodman) [reply sycl/source/detail/platform_impl.cpp]: ok

File: sycl/source/detail/platform_impl.hpp [hunk 16]
```diff
@@ -180,12 +182,16 @@ class platform_impl {
...
 
+  context getDefaultContext();
+
 private:
   bool MHostPlatform = false;
   RT::PiPlatform MPlatform = 0;
   std::shared_ptr<plugin> MPlugin;
   std::vector<std::weak_ptr<device_impl>> MDeviceCache;
   std::mutex MDeviceMapMutex;
+  std::shared_ptr<context> MDefaultContext;
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.hpp:193]: `context` is a wrapper around `std::shared_ptr<context_impl>`. So, suggest storing shared_ptr to context_impl directly here. And create a `context` from it by calling `detail::createSyclObjFromImpl<context>` like in source/backend.cpp:76 .

File: sycl/doc/extensions/PlatformContext/PlatformContext.adoc [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+
+The platform class gains one new method:
+
+[cols="^60a,40"]
```
Comment (romanovvlad) [inline sycl/doc/extensions/PlatformContext/PlatformContext.adoc:15]: It seems now to be complaint with SYCL2020 the extension should define a feature macro and prefix public methods with `ext_`.

Comment (jbrodman) [reply sycl/doc/extensions/PlatformContext/PlatformContext.adoc:15]: ok.

**Review by alexbatashev — DISMISSED** [review]
LGTM

**Review by mkinsner — DISMISSED** [review]

**Comment by jbrodman** [comment]
I don't know what the deal is with testing.  Can someone please tell me why it's unhappy?

**Comment by bader** [comment]
It looks the like there is a deadlock somewhere because all tests are failing due to exceeded time limit.

**Comment by jbrodman** [comment]
I have no idea why that would be the case.

File: sycl/source/detail/platform_impl.cpp [hunk 18]
```diff
@@ -9,6 +9,7 @@
 #include <CL/sycl/device.hpp>
 #include <detail/allowlist.hpp>
 #include <detail/config.hpp>
+#include <detail/context_impl.hpp>
```
Comment (romanovvlad) [inline sycl/source/detail/platform_impl.cpp:12]: Will remove

File: sycl/source/detail/queue_impl.hpp [hunk 19]
```diff
@@ -49,6 +49,18 @@ enum QueueOrder { Ordered, OOO };
 
 class queue_impl {
 public:
+  // \return a default context for the platform if it includes the device
+  // passed, a new context otherwise.
+  static ContextImplPtr getDefaultOrNew(const DeviceImplPtr &Device) {
```
Comment (romanovvlad) [inline sycl/source/detail/queue_impl.hpp:54]: We cannot use the default context(which is bound to root devices) to create a queue for a sub-device, so here we create a new context if for some reason the default context doesn't include target device.
We could bound all sub-devices to the same default context as well, but it might degrade performance.
Also, there are a lot of ways to create sub-devices from a root device, so it's not clear which sub-devices should be included.

Comment (romanovvlad) [reply sycl/source/detail/queue_impl.hpp:54]: Tagging @jbrodman

File: sycl/doc/EnvironmentVariables.md [hunk 20]
```diff
@@ -47,6 +47,7 @@ subject to change. Do not rely on these variables in production code.
 | `SYCL_CACHE_THRESHOLD` | Positive integer | Cache eviction threshold in days (default value is 7 for 1 week). Set to 0 for disabling time-based cache eviction. |
 | `SYCL_CACHE_MIN_DEVICE_IMAGE_SIZE` | Positive integer | Minimum size of device code image in bytes which is reasonable to cache on disk because disk access operation may take more time than do JIT compilation for it. Default value is 0 to cache all images. |
 | `SYCL_CACHE_MAX_DEVICE_IMAGE_SIZE` | Positive integer | Maximum size of device image in bytes which is cached. Too big kernels may overload disk too fast. Default value is 1 GB. |
+| `SYCL_ENABLE_DEFAULT_CONTEXTS` | '1' or '0' | Enable ('1') or disable ('0') creation of default contexts in SYCL runtime. Enabled by default on Linux and disabled on Windows. |
```
Comment (bader) [inline sycl/doc/EnvironmentVariables.md:50]: I guess we already have "default contexts" in SYCL runtime, but they work differently. I suggest we clarify that it enables creation of "platform default context" and describe how it's behavior is different from regular "default context". Otherwise there might be a confusion. What do you think?

File: sycl/unittests/misc/DefaultContext.cpp [hunk 21]
```diff
@@ -34,3 +49,21 @@ TEST(DefaultContextTest, DefaultContextTest) {
   ASSERT_EQ(Dev1.get_platform().ext_oneapi_get_default_context(),
             Dev2.get_platform().ext_oneapi_get_default_context());
 }
+
+TEST(DefaultContextTest, DefaultContextCanBeDisabled) {
+  set_env(EnableDefaultContextsName, "0");
+
```
Comment (romanovvlad) [inline sycl/unittests/misc/DefaultContext.cpp:55]: Shouldn't we reset config after setting this env var?

File: sycl/doc/EnvironmentVariables.md (hunk 20 above)
Comment (romanovvlad) [reply sycl/doc/EnvironmentVariables.md:50]: Could you please clarify what "default contexts" we already have?

File: sycl/unittests/misc/DefaultContext.cpp (hunk 21 above)
Comment (alexbatashev) [reply sycl/unittests/misc/DefaultContext.cpp:55]: I think I broke caching mechanism for SYCL config. I'll fix it.

**Comment by alexbatashev** [comment]
Tests should be fixed after https://github.com/intel/llvm-test-suite/pull/433

**Comment by alexbatashev** [comment]
SYCL :: Plugin/level_zero_dynamic_batch_test.cpp is failing
There's a request to disable that test: https://github.com/intel/llvm-test-suite/pull/453

**Review by bader — APPROVED** [review]

File: sycl/source/detail/platform_impl.cpp [hunk 22]
```diff
@@ -17,6 +17,7 @@
 
 #include <algorithm>
 #include <cstring>
+#include <mutex>
```
Comment (s-kanaev) [inline sycl/source/detail/platform_impl.cpp:20]: NIT: Is this include needed in this file?

**Review by s-kanaev — DISMISSED** [review]
LGTM

File: sycl/source/detail/platform_impl.cpp (hunk 22 above)
Comment (bader) [reply sycl/source/detail/platform_impl.cpp:20]: Line 40.

File: sycl/source/detail/queue_impl.hpp [hunk 23]
```diff
@@ -50,6 +51,22 @@ enum QueueOrder { Ordered, OOO };
 
 class queue_impl {
 public:
+  // \return a default context for the platform if it includes the device
+  // passed and defaualt contexts are enabled, a new context otherwise.
```
Comment (bader) [inline sycl/source/detail/queue_impl.hpp:55]:
```suggestion
  // passed and default contexts are enabled, a new context otherwise.
```

**Review by bader — APPROVED** [review]

**Review by romanovvlad — APPROVED** [review]

**Review by s-kanaev — APPROVED** [review]
LGTM

**Comment by alexbatashev** [comment]
@bader @romanovvlad do you know what's wrong with cuda build? It's been stuck for over a day, but I don't see a huge queue of other PRs.

**Comment by bader** [comment]
> do you know what's wrong with cuda build?

No.

**Comment by alexbatashev** [comment]
The job status is still not updated. I reported this to tooling team. The difference between last commit and the previous one is just one comment, and the previous commit has this job passed. Can we assume that tests pass?

**Comment by bader** [comment]
@romanovvlad, @alexbatashev, unittests do not pass on GitHub machines. 
https://github.com/intel/llvm/runs/3564775299 
Please, fix ASAP.
